### PR TITLE
Records API performance improvements

### LINKED ISSUE #1763: Improve `/records/` API performance
## Problem
Records API takes around 5 seconds to respond, sometimes taking even longer. It mostly happens when there is a foreign key column 

## Proposed solution
There are two possible fixes that could improve performance
- Compile Record summary queries and store them in the database instead of recursively constructing them during the request.
- Make the `get_column_name_id_bidirectional_map` prefetch fetch column properties and related objects

### PR DESCRIPTION
**Pull request by silentninja**
Fixes #1763 

Uses cached metadata and some Django prefetches to improve the performance of the `/records/` API. 

**Technical Details**

The performance improvements made by the PR are as follows
- SqlAlchemy queries were reduced from 660 to 141 queries
- Django queries were reduced from 194 queries to 91

## Checklist
 
- [x] My pull request has a descriptive title (not a vague title like `Update
  index.md`).
- [ ] My pull request targets the `master` branch of the repository
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [ ] I added tests for the changes I made (if applicable).
- [ ] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no
  visible errors.

[best_practices]:https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

**Comment by silentninja** [comment]
I am marking this PR as draft as it depends on #1851. While the changes by themselves don't depend on each other, #1851 also improves the `/records/` API performance and I wanted to prevent any conflicting work.

**Review by mathemancer — COMMENTED** [review]
This adds state and complexity. Do you have any metrics to offer showing that the benefits outweigh these costs?

File: mathesar/api/db/viewsets/records.py [hunk 1]
```diff
@@ -49,7 +49,6 @@ def list(self, request, table_pk=None):
         column_names_to_ids = table.get_column_name_id_bidirectional_map()
         column_ids_to_names = column_names_to_ids.inverse
         if filter_unprocessed:
-            table = get_table_or_404(table_pk)
```
Comment (mathemancer) [inline mathesar/api/db/viewsets/records.py:52]: What's going on here? Was this cruft left over from something?

File: mathesar/api/pagination.py [hunk 2]
```diff
@@ -78,7 +78,7 @@ def paginate_queryset(
         preview_metadata = None
         # Only tables have columns on the Service layer that hold data necessary for preview template.
         if isinstance(table, Table):
-            columns_query = table.columns.all()
+            columns_query = Column.objects.filter(table_id=table.id).select_related('table__schema__database').prefetch('name')
```
Comment (mathemancer) [inline mathesar/api/pagination.py:81]: How unfortunate that `table.columns` doesn't work here.

**Review by mathemancer — APPROVED** [review]
Okay, I don't see any problems here. I just had a couple of non-blocking questions.

File: mathesar/api/db/viewsets/records.py (hunk 1 above)
Comment (silentninja) [reply mathesar/api/db/viewsets/records.py:52]: Yes

File: mathesar/api/pagination.py (hunk 2 above)
Comment (silentninja) [reply mathesar/api/pagination.py:81]: I was annoyed by that too and tried looking for a better solution, but the [Django suggested solution](https://docs.djangoproject.com/en/1.10/releases/1.10/#manager-use-for-related-fields-and-inheritance-changes) doesn't seem to be working, I will make this refactor sometime later

**Pull request merged**